api/s2n.h

```diff
@@ -1960,6 +1960,32 @@ S2N_API extern int s2n_connection_free(struct s2n_connection *conn);
  */
 S2N_API extern int s2n_shutdown(struct s2n_connection *conn, s2n_blocked_status *blocked);
 
+typedef enum {
+    /* The write side of the connection is closed.
+     * IO methods that write, like s2n_send(), cannot be called. */
+    S2N_WRITE_CLOSED,
+    /* The read side of the connection is closed.
+     * IO methods that read, like s2n_recv(), cannot be called. */
+    S2N_READ_CLOSED,
+    /* The read and write sides of the connection are both closed.
+     * No IO methods can be called.
+     */
+    S2N_FULL_CLOSED,
+} s2n_closed_type;
+
+/**
+ * Reports whether or not the connection is closed.
+ *
+ * TLS1.2 only considers a connection closed or not closed, but TLS1.3 supports
+ * "half-close" behavior where the read and write sides of a connection can be
+ * closed separately.
+ *
+ * @param conn A pointer to the s2n_connection object
+ * @param side The type of closure being queried
+ * @returns true if closed, false otherwise
+ */
+S2N_API extern bool s2n_connection_is_closed(struct s2n_connection *conn, s2n_closed_type side);
```

Comment:
I was just curious what the use case is for this API? Wouldn't you know if you're able to send/receive or not just by attempting it?

Reply:
Yes, but you'd have to attempt it ;) If you don't have anything you want to send/receive at that moment, you couldn't check for a closed connection.

Reply:
A bit of a nit: I'm wondering if it would make for a clearer API to ask if the connection is open rather than if it's closed. I think talking about the "positive" case can lead to a bit clearer code in the application and fewer double negatives.

```c
typedef enum {
    S2N_IO_WRITABLE,
    S2N_IO_READABLE,
    S2N_IO_FULL_DUPLEX,
    S2N_IO_CLOSED,
} s2n_io_status;

S2N_API extern bool s2n_connection_io_status(struct s2n_connection *conn, s2n_io_status query);
```

which results in the following application code:

```c
if (s2n_connection_io_status(conn, S2N_IO_WRITABLE)) {
    // The connection is writable so i can call send!
    s2n_send("Hello World", sizeof("Hello World"), &blocked);
}
```

rather than:

```c
if (!s2n_connection_is_closed(conn, S2N_WRITE_CLOSED)) {
    // The connection is not write-closed so I can call send!
    s2n_send("Hello World", sizeof("Hello World"), &blocked);
}
```

Also, phrasing it as a "status" rather than a `is_open`/`is_closed` means you can ask if it's fully closed or fully open in a single call.

Reply:
Love the suggestion, definitely taking it.

I'm thinking I may also postpone making this API public until another PR. This one is a bit big for permanent API changes.